pinot-plugins/pinot-input-format/pinot-csv/src/main/java/org/apache/pinot/plugin/inputformat/csv/CSVRecordReader.java

```diff
@@ -60,102 +61,119 @@ public class CSVRecordReader implements RecordReader {
-      }
+  private static CSVFormat.Builder formatBuilder(CSVRecordReaderConfig config) {
+    final CSVFormat.Builder builder = baseCsvFormat(config).builder().setDelimiter(config.getDelimiter()).setHeader()
+        .setSkipHeaderRecord(config.isSkipHeader()).setCommentMarker(config.getCommentMarker())
+        .setEscape(config.getEscapeCharacter()).setIgnoreEmptyLines(config.isIgnoreEmptyLines())
+        .setIgnoreSurroundingSpaces(config.isIgnoreSurroundingSpaces()).setQuote(config.getQuoteCharacter());
+
+    optional(config.getQuoteMode()).map(QuoteMode::valueOf).ifPresent(builder::setQuoteMode);
+    optional(config.getRecordSeparator()).ifPresent(builder::setRecordSeparator);
+    optional(config.getNullStringValue()).ifPresent(builder::setNullString);
+    return builder;
+  }
 
-      String nullString = config.getNullStringValue();
-      if (nullString != null) {
-        _format = _format.builder().setNullString(nullString).build();
+  private static Map<String, Integer> parseLineAsHeader(CSVFormat format, String line)
+      throws IOException {
+    try (StringReader stringReader = new StringReader(line)) {
+      try (CSVParser parser = format.parse(stringReader)) {
+        return parser.getHeaderMap();
       }
+    }
+  }
+
+  private static Character getMultiValueDelimiter(CSVRecordReaderConfig config) {
+    if (config == null) {
+      return CSVRecordReaderConfig.DEFAULT_MULTI_VALUE_DELIMITER;
+    } else if (config.isMultiValueDelimiterEnabled()) {
+      return config.getMultiValueDelimiter();
+    }
+    return null;
+  }
 
+  private static boolean useLineIterator(CSVRecordReaderConfig config) {
+    return config != null && config.isSkipUnParseableLines();
+  }
+
+  @Override
+  public void init(File dataFile, @Nullable Set<String> fieldsToRead, @Nullable RecordReaderConfig recordReaderConfig)
+      throws IOException {
+    _dataFile = dataFile;
+    _config = (CSVRecordReaderConfig) recordReaderConfig;
+    if (_config == null) {
+      _format = defaultFormat();
+    } else {
+      _isHeaderProvided = _config.getHeader() != null;
+      final CSVFormat.Builder builder = formatBuilder(_config);
       if (_isHeaderProvided) {
-        _headerMap = parseLineAsHeader(config.getHeader());
-        _format = _format.builder().setHeader(_headerMap.keySet().toArray(new String[0])).build();
-        if (!_useLineIterator) {
-          validateHeaderForDelimiter(delimiter, config.getHeader(), _format);
-        }
+        // use an intermediate format to parse the header line. It still needs to be updated later
+        _headerMap = parseLineAsHeader(builder.build(), _config.getHeader());
+        builder.setHeader(_headerMap.keySet().toArray(new String[0]));
       }
+      _format = builder.build();
 
-      if (config.isMultiValueDelimiterEnabled()) {
-        multiValueDelimiter = config.getMultiValueDelimiter();
+      if (_isHeaderProvided) {
```

Comment:
The isMultiValueDelimiterEnable check is now moved to getMultiValueDelimiter and we are setting this param in newCsvRecordExtractorConfig, right?

Reply:
yeah. I didn't really see a lot of benefit maintain a cached value of the config unless it is being iterated on or accessed frequently.

The `CSVRecordExtractor` uses this to capture multi value fields. `CSVRecordReader` doesn't use it.